File: mek_stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/RefundMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.Refund
import RefundApi
import RefundStatusApi
import mek.stripeterminal.toHashMap

fun Refund.toApi(): RefundApi {
    return RefundApi(
        id = id,
        amount = amount,
        chargeId = chargeId,
        createdInMilliseconds = created?.times(1000),
        currency = currency,
        metadata = metadata?.toHashMap(),
        reason = reason,
        status = when (status) {
            "succeeded" -> RefundStatusApi.SUCCEEDED
            "pending" -> RefundStatusApi.PENDING
            "failed" -> RefundStatusApi.FAILED
            else -> null
        },
        paymentMethodDetails = paymentMethodDetails?.toApi(),
        failureReason = failureReason
    )
}

File: mek_stripe_terminal/ios/mek_stripe_terminal/Sources/mek_stripe_terminal/Mappings/RefundMappings.swift
import Foundation
import StripeTerminal

extension Refund {
    func toApi() -> RefundApi {
        return RefundApi(
            id: stripeId,
            amount: amount.toInt64(),
            chargeId: chargeId,
            createdInMilliseconds: created.toMillisecondsSinceEpoch(),
            currency: currency,
            metadata: metadata,
            reason: reason,
            status: status.toApi(),
            paymentMethodDetails: paymentMethodDetails?.toApi(),
            failureReason: failureReason
        )
    }
}

extension RefundStatus {
    func toApi() -> RefundStatusApi? {
        switch self {
        case .succeeded:
            return .succeeded
        case .pending:
            return .pending
        case .failed:
            return .failed
        case .unknown:
            return nil
        @unknown default:
            fatalError()
        }
    }
}

extension PaymentMethodDetails {
    func toApi() -> PaymentMethodDetailsApi {
        return PaymentMethodDetailsApi(
            cardPresent: cardPresent?.toApi(),
            interactPresent: interacPresent?.toApi()
        )
    }
}
